File: 백준/Gold/1922. 네트워크 연결/네트워크 연결.cc
#include<iostream>
#include<algorithm>
#include<vector>
#include<tuple>
#define FAST_IO cin.tie(0), cout.tie(0), ios_base::sync_with_stdio(0)
using namespace std;

int n, m;
vector<tuple<int, int, int>> edges;
vector<int> par;

int Find(int x) {
	return x == par[x] ? x : par[x] = Find(par[x]);
}

void Union(int a, int b) {
	a = Find(a);
	b = Find(b);

	if (a < b) par[b] = a;
	else par[a] = b;
}

int kruskal() {

	int mst = 0;
	int cnt = 0;

	for (auto cur : edges) {
		int cost = get<0>(cur);
		int u = get<1>(cur);
		int v = get<2>(cur);

		if (Find(u) == Find(v)) continue;
		Union(u, v);
		cnt++;
		mst += cost;

		if (cnt == n - 1) break;
	}

	return mst;

}

int main(void) {
	FAST_IO;

	cin >> n >> m;

	par.resize(n + 1);
	for (int i = 0; i < n + 1; i++) par[i] = i;
	

	for (int i = 0; i < m; i++) {
		int u, v, c;
		cin >> u >> v >> c;
		edges.push_back({c, u, v});
		edges.push_back({c, v, u});
	}

	sort(edges.begin(), edges.end());

	int mst_sum = kruskal();

	cout << mst_sum;

	return 0;

	


}

File: 백준/Gold/1922. 네트워크 연결/네트워크 연결.py
import sys

sys.setrecursionlimit(10**6)

input = sys.stdin.readline

def find(x):
    if par[x] == x :
        return x
    else:
        par[x] = find(par[x])
        return par[x]
   

def union(a, b):
    a = find(a)
    b = find(b)
    if a < b :
        par[b] = a
    else:
        par[a] = b

def kruskal():
    ret = 0
    cnt = 0

    for c, u, v in edges:
        if find(u) == find(v):
            continue
        if cnt == n - 1:
            break

        union(u, v)
        ret += c
        cnt += 1

    return ret


n = int(input())
m = int(input())
edges = []
mst = 0
par = [i for i in range(n+1)]

for _ in range(m):
    a, b, c = map(int, input().split())
    edges.append((c, a, b))

edges.sort()

print(kruskal())